### ISSUE TITLE
Benchmark real-world use cases

### ISSUE THREAD

**Issue by rtsao** [hellofresh/css-in-js-perf-tests#18]
I think having some large fixture of real-ish styles would be a useful benchmark to better reflect real-world use cases. That way there's an appropriate blend of duplicate and unique styles that would more accurately reflect actual usage.

**Comment by PepijnSenders** [comment]
Hey @rtsao, I was thinking the same, any ideas?

**Comment by PhiLhoSoft** [comment]
Yes, I am working on a complex-test test case, precisely for the same reason.
I thought about making a header, footer, main area with an aside, all laid out with flexbox and with a max-width media query to change the layout. This allows to test, among other things, if auto-prefixing is done by libraries or not.
The main area would have a list with several items using the same style (testing effective use of classes), with a :last-child pseudo-selector.
This kind of things, pushing a bit more these libraries to their limits... 😄

**Comment by PhiLhoSoft** [comment]
Another thing to explore is class composition. I see CSJS allows this, I don't know if Styletron allows this (or if it makes sense for it), idem to scope-styles. To take three examples from the same origin... 😉 

BTW, @rtsao, maybe it would be interesting in each readme of your projects to tell their usage / interest at least relative to each other... 😄

**Comment by pygy** [comment]
Regarding composition, j2c allows to compose source objects using arrays, so `[{foo:5}, {bar:6}]` becomes `{foo:5, bar:6}`. It works at any level of nesting, at the selector, declaration and value levels (and you can nest arrays).

`j2c` doesn't do any deduplication. Preserving the source order in the output is part of the design goals. Gzip deals well with exact dupes (although even after compression, Fela and Styletron still produce consistently smaller output).

I'm currently working on an auto-prefixer plugin that uses Autoprefixer/PostCSS on the server and PrefixFree on the client. I'll keep you posted.

**Comment by PhiLhoSoft** [comment]
Deduplication is nice but not essential if you write correctly your CSS... This bench gives interesting information, but perf is not the utmost goal (in lot of cases), ease of use, flexibility, etc. are important too. That's why I do this complex test case, to explore more the capabilities of the libraries.

I have yet to see this PrefixFree project (a link for the lazy?), most libraries seem to use https://github.com/rofrischmann/inline-style-prefixer instead, which apparently work on client and server. Annoyance: on client it targets the current browser only, which is nice, using the user-agent, which is less nice (but auto-detection in CSS is hard); but on the server, it prefixes for all browsers, no version range selection as in Autoprefixer.

**Comment by PhiLhoSoft** [comment]
Thinking more about it, deduplication can be annoying sometime.
At this point, I prefer the method of keeping the original name with suffix (or prefix) and not deduplicating (but removing unused classes is still good), because it will probably make debugging and experimenting in DevTools easier (something I plan to test later).
Eg. you have two unrelated classes that has same content. If deduplicated and if you try to change an element using one of these classes (eg. changing a margin), you will get the effect on the unrelated elements as well... :-/
These are points to take in account in the process of choosing a library...

**Comment by pygy** [comment]
I mentioned a post that disappeared in another thread, actually I mixed things up and didn't notice you had posted here too...

https://github.com/LeaVerou/prefixfree is an old (2011, not on NPM) automatic prefixer tool that doesn't use browser sniffing. It is meant to be included as is in the source to patch stylesheets. 

Out of the box, it has several passes per sheet (one for selectors, one for at rules, one for properties, etc...) using complex regexps that are however oblivious to strings. This is both inefficient and a bit sloppy for my use case (a proper handling of strings would be slower still, though).

I've adapted the feature detection code for a post-processing `j2c` filter, using recast to move code automatically and safely. [Here](https://github.com/j2css/j2c/blob/v1_preview/plugins/prefix-browser/src/extracted/properties.js)'s the code that determines what prefix to use and what properties need a prefix. That's all Lea's work, extracted into a file that can interoperate (the original is all locked up in closures).

What this doesn't do, though, critically, is handle the flexbox mess. I'll have to add a way to detect and convert it.

**Comment by pygy** [comment]
@PhiLhoSoft Thinking more about a "real world" benchmark, I'd suggest taking a page of the JSS benchmark, and measure the time it takes to get to `document.onload`, when doing a pure client side render and maybe a hydration.

Otherwise, you don't measure the time it takes to parse, load and setup the libraries into account (or you do it partially, as when you load the JSS presets in the benchmark loop.

Also, in real use cases, loading and configuring the libs is something that happens once or a few times (for the config part) at page load time, and it is unlikely to get the Crankshaft/Turobfan/$FastJIT treatment, so it makes little sense to put it in a benchmark loop where it is likely to be optimized.

**Comment by kof** [comment]
It was, I removed it some time ago, because it was just a manual test. It needs to be statistically significant.

**Comment by pygy** [comment]
We could build a harness that loads tests like that in an iframe. It would allow to do repeated runs of a full page load to filter out the noise and get accurate measurements.

**Comment by kof** [comment]
> We could build a harness that loads tests like that in an iframe. It would allow to do repeated runs of a full page load to filter out the noise and get accurate measurements.

I haven't seen this done this way somewhere else. It might be a lot of side effects involved.

**Comment by pygy** [comment]
Indeed, we'd have to be careful... You'd want the code to be loaded from a hot cache, but then hope that the JS engine really reparses and reloads the JS... IIRC some engines keep a code cache for code shared between frames. This may or may not be relevant here, I don't remember enough details.

**Comment by PepijnSenders** [comment]
I agree with @kof as well, I'm not sure how deep this goes, but can't we use some isolated environment of the chromium engine to isolate as many side effects as possible?

**Comment by kof** [comment]
Maybe we can add just different test cases like rendering 200kb of styles at once with and without full dom render.

**Comment by pygy** [comment]
@PepijnSenders It could be a solution... I'm only familiar with PhantomJS which usually lags behind Chrome...

@kof Having a DOM is a must, I'm curious about the time it takes to process the many classes generated by Fela and StyleTron (whose approach is so far most impressive).

**Comment by rtsao** [comment]
FWIW I wrote some client-side benchmarks powered by Chrome/[Lighthouse](https://github.com/GoogleChrome/lighthouse)/[pwmetrics](https://github.com/paulirish/pwmetrics) in the Styletron repo: https://github.com/rtsao/styletron/tree/master/packages/benchmarks

**Comment by PepijnSenders** [comment]
Ah Lighthouse, great idea!

**Comment by PepijnSenders** [comment]
Hm is it me though or is pwmetrics a bit broken? 😛

**Comment by rtsao** [comment]
@PepijnSenders Last time I checked they were working... you need to run the static file server (`npm run serve-static`), did you forget that?

I also think we are conflating two issues here:

1) Benchmarking client-side render and/or rehydration performance
2) Benchmarking real-world style use cases.

To solve the second point, I propose creating a small library that can generate `n` different styles with tweakable parameters that control the following metrics:

1) Total number of style objects
2) Percentage of duplicate style objects (0 - 100%)
3) Mean declarations per style object (1 - `n`)
4) Variation of declarations per style object
5) Percentage of duplicate declarations (0 - 100%)
6) and more stuff

**Comment by pygy** [comment]
@rtsao Ideally, the parameters chosen for the tests should reflect the distribution of these parameters in real apps/sites.

**Comment by rtsao** [comment]
Agreed. Some resources on this:

http://cssstats.com/
https://github.com/pankajparashar/atomize-io
http://www.stylestats.org/

I also did some investigation into this (results here: https://github.com/rtsao/styletron/issues/46#issuecomment-266673724)

**Comment by pygy** [comment]
That's a good starting point.

None of these tools analyse styles generated with JS, though, but I don't think it would change much in the stats anyway.

**Comment by PhiLhoSoft** [comment]
As a reminder, this project, so far, only tests performances on the server side. I don't know if tests on the client side, needing perhaps quite some different code, fits in this project or should be done in a different one.

Personally, I find performance° to be an interesting point to take in account, but not necessarily essential.
Or, rather, I also value ease of use, flexibility, clean generated code, ease of maintenance, including debugging and experimenting with dev tools, and so on...

° And there are two different perf measures, apparently: the one Aphrodite (Khan Academy) sees as important, the rendering after first load, indeed important for a regular web site, and one where we measure speed of updates, more important for SPAs where initial load time is a minor concern. I saw a page (can't find it back) where a full page of components (system stats?) are updated at high speed, for example.

To test SPA dev, I plan to do a side project of a semi-real application, roughly based on my current [complex-test-case](https://github.com/PhiLhoSoft/css-in-js-perf-tests/tree/complex-test-case), for rendering on the client side (can be quite different with some libraries) and testing other stuff like dynamic update of styles (change style of button based on state of a form, this kind of stuff).
I would use some lightweight, relatively neutral view library, like Inferno perhaps.
And it can be used as a base for automated tests to benchmark the speed... @rtsao's code can be useful there... 😄

**Comment by kof** [comment]
> Personally, I find performance° to be an interesting point to take in account, but not necessarily essential.
Or, rather, I also value ease of use, flexibility, clean generated code, ease of maintenance, including debugging and experimenting with dev tools, and so on...

Very valid point. Maybe a disclaimer like that should be put in a good visible place in the readme.

**Comment by kof** [comment]
A good start here would be to randomize the props and values, so that the test becomes a bit closer to reality. I guess some libs perform insanely fast because they basically cached the results and do nothing.

**Comment by PepijnSenders** [comment]
Can't we do some compares of the generated styles from some popular websites like facebook or amazon. Just a simple mock-up of a simple-ish page.

**Comment by kof** [comment]
That would be ideal, though you would need to convert it to every libs syntax, which have some differences each.

**Comment by PepijnSenders** [comment]
Most libaries work with some kind of className generation, we could build an interface for that.

**Comment by PhiLhoSoft** [comment]
"*I guess some libs perform insanely fast because they basically cached the results and do nothing.*"
The fastest are, by a large margin, Fela and Styletron, using the same method (one class per rule).
And they use a fresh instance per test. Other libraries have a flush / reset that is supposed to take care of caches (in theory).

That said, I stumbled upon a very strange issue that makes me doubt of test isolation: if I add auto-prefixing to CXS, for example, via `inline-style-prefixer/static`, several other libraries (fela, cxs-optimized, free-style and j2c) output auto-prefixed code, without changing anything to the code of their case! Confused...

**Comment by kof** [comment]
I ment the json syntax is different in some cases.

**Comment by PhiLhoSoft** [comment]
@kof yes, as I go deeper in some features, and add other libraries, I had to make some features depending on options. Some libraries want their class names prefixed with a dot, others want & before pseudo-classes, JSS treats fallbacks differently than other libraries (not an array of values), globals (html, body, etc.) are not treated the same, and so on. It becomes hard to have a same style for everybody, even if they use a very similar syntax.

**Comment by pygy** [comment]
@kof, re. randomization: On the server side, or when re-rendering components, the caching done by Fela and Styletron is a feature, and a real advantage that should not be masked either...

Re. generating test cases from real CSS, http://j2c.py.gy/import.html turns most CSS into JSON that `j2c` understands. (It also supports SCSS-like nested rules, by accident, but may not play well with CSS variables).

**Comment by kof** [comment]
>  the caching done by Fela and Styletron is a feature, 

Sure it is, though using the same props and values all the time makes it cache unrealistically much. I can add a cache to JSS in 5 LOC and after the first render it will do nothing. So what would we compare then?

**Comment by pygy** [comment]
I think that was covered in @rtsao's [suggestion above](https://github.com/hellofresh/css-in-js-perf-tests/issues/18#issuecomment-268900766).

Caching per source object would mean their engines would be oblivious to source object mutations. What they most likely do (I didn't check) is keep a `(atRule?, pseudoSelector?, property, value)` => `atomicClass`  mapping and check it while iterating on the source. Likewise, they likely cache a `sortedClassList` => `css` mapping. At least that's how I would do it...

So each test iterations must either do, per key-value pair, string concatenation or several object lookups, to check the cache.

The stats that @rtsao posted show that the atomization of the styles of real sites shrinks the resulting CSS. That means that there were duplicate styles in the source that would benefit from the caching Fela and Styletron do, even when loading a single page.

The data we use for the "real world" benchmark should reflect these duplications. We may sample properties randomly, with a controlled amount of duplication that reflects real sites.

Note that Like JSS, `j2c` doesn't do any caching, I'm not trying to skew the benchmark in its favor.

**Comment by kof** [comment]
Yeah, I guess we could generate samples that are close to realistic, by calculating average metrics from some popular sites and using the metrics to generate the samples.

**Comment by axemclion** [comment]
The test cases are indeed skewed. I just ran tests to see how well each of the frameworks do while rendering on a browser - and here are the results 

<img width="716" alt="image" src="https://cloud.githubusercontent.com/assets/17868/22237254/6ac6697c-e1bf-11e6-88f7-6457d913a0ae.png">

Some frameworks do have different characteristics based on specific test cases.
